x-pack/plugins/index_management/public/application/sections/home/data_stream_list/edit_data_retention_modal/edit_data_retention_modal.tsx

```diff
@@ -0,0 +1,327 @@
+      {
+        defaultMessage: 'Time unit',
+      }
+    ),
+  },
+  infiniteRetentionPeriod: {
+    type: FIELD_TYPES.TOGGLE,
+    defaultValue: false,
+    label: i18n.translate(
+      'xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.infiniteRetentionPeriodField',
+      {
+        defaultMessage: 'Never delete data',
+      }
+    ),
+  },
+};
+
+export const EditDataRetentionModal: React.FunctionComponent<Props> = ({
+  dataRetention,
+  dataStreamName,
+  configuredByILM,
+  onClose,
+}) => {
+  const { size, unit } = splitSizeAndUnits(dataRetention);
+  const {
+    services: { notificationService },
+  } = useAppContext();
+
+  const { form } = useForm({
+    defaultValue: {
+      dataRetention: size,
+      timeUnit: unit || 'd',
+      infiniteRetentionPeriod: !dataRetention,
+    },
+    schema: configurationFormSchema,
+    id: 'editDataRetentionForm',
+  });
+  const [formData] = useFormData({ form });
+  const isDirty = useFormIsModified({ form });
+
+  const onSubmitForm = async () => {
+    const { isValid, data } = await form.submit();
+
+    if (!isValid) {
+      return;
+    }
+
+    return updateDataRetention(dataStreamName, data).then(({ data: responseData, error }) => {
+      if (responseData) {
+        const successMessage = i18n.translate(
+          'xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.successDataRetentionNotification',
+          { defaultMessage: 'Data retention updated' }
+        );
+        notificationService.showSuccessToast(successMessage);
+
+        return onClose({ hasUpdatedDataRetention: true });
+      }
+
+      if (error) {
+        const errorMessage = i18n.translate(
+          'xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.errorDataRetentionNotification',
+          {
+            defaultMessage: "Error updating data retention: '{error}'",
+            values: { error: error.message },
+          }
+        );
+        notificationService.showDangerToast(errorMessage);
+      }
+
+      onClose();
+    });
+  };
+
+  return (
+    <EuiModal onClose={() => onClose()} data-test-subj="editDataRetentionModal">
+      <Form form={form} data-test-subj="editDataRetentionForm">
+        <EuiModalHeader>
+          <EuiModalHeaderTitle>
+            <FormattedMessage
+              id="xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.modalTitleText"
+              defaultMessage="Edit data retention"
+            />
+          </EuiModalHeaderTitle>
+        </EuiModalHeader>
+
+        <EuiModalBody>
+          {configuredByILM && (
+            <>
+              <EuiCallOut
+                color="warning"
+                iconType="warning"
+                data-test-subj="configuredByILMWarning"
+                title={i18n.translate(
+                  'xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.configuredByILMWarningTitle',
+                  {
+                    defaultMessage: 'Override ILM data retention',
```

Comment:
```suggestion
                    defaultMessage: 'Current data retention conflicts with ILM policy ',
```
Feel free to ignore or adjust, see my comment below on the callout body